src/main/java/com/amazon/ion/impl/bin/IonManagedWriter_1_1.kt

```diff
@@ -135,45 +145,299 @@ internal class IonManagedWriter_1_1(
         sid = newSymbols[text]
         if (sid != null) return sid
         // Add to the to-be-appended symbols
-        sid = priorMaxId + newSymbols.size + 1
+        sid = symbolTable.size + newSymbols.size + 1
         newSymbols[text] = sid
         return sid
     }
 
-    /** Writes a Local Symbol Table for the current encoding context, and updates the prior context. */
-    private fun writeSymbolTable() {
-        if (newSymbols.isEmpty()) return
-        val useSid = options.internEncodingDirectiveSymbols
+    private fun internMacro(macro: Macro): Int {
+        // Check the current macro table
+        var id = macroTable[macro]
+        if (id != null) return id
+        // Check the to-be-appended macros
+        id = newMacros[macro]
+        if (id != null) return id
+        // Add to the to-be-appended symbols
```

Comment:
Should this check the `newMacros` first so new definitions shadow the old ones?

Reply:
No—at least not here—because this uses deep equality of the macro signature and body. (In practice, though, it should be short-circuited by a `this === other` check most of the time.)

However, your question is making me rethink the `addMacro(...): MacroRef` function because `MacroRef` could outlive an encoding context with no good way to determine whether it's still valid.